docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -424,299 +442,321 @@ In Geo-Partitioning, the table must be splitted into partitions, and each partit
+2. It will probably be used in locality-aware features in the future, such as follower-read. Follower-read is always used when TiDB wants to read the nearest replica to reduce multi-region latency. In some distributed databases, it’s implemented by labelling data nodes and selecting the nearest replica according to the labels.
+3. Local transactions need to know the binding relationship between Raft leader and region, which is also defined by placement rules.
+4. Once a rule is defined on a table, all the subsequent partitions added to the table should also inherit the rule. So the `ADD PARTITION` operation should query the rules on the table. The same is true for creating tables and indices.
+5. The `SHOW PLACEMENT` statement should output the placement rules correctly.
 
-All placement rules can be queried through statements.
+As placement rules will be queried in case 1, 2 and 3, low latency must be guaranteed. As discussed in section "Storing placement rules", PD is the source of truth. To lower the latency, the only way is caching the placement rules in TiDB.
 
-### System table
+Since the cache is created, there must be a way to validate it. Different from region cache, placement rules cache can only be validated each time from PD. There are some ways to work around:
 
-A new system table `information_schema.placement_rules` is added to view all placement rules. The table contains such columns:
+- Update the schema version once a placement rule is changed, just like other DDL. PD broadcasts the latest schema version to all the TiDB instances, and then TiDB instances fetch the newest placement rules from PD. There will be a slight delay for queries before reading the latest placement rules. The side effect is that more transactions will retry since the schema version is changed.
+- TiDB queries placement rules from PD periodly. The delay is controllable but not eliminable.
+- Once a placement rule is changed, PD broadcasts it to all the TiDB instances. In this approach, the schema version is not involved, so transactions are not affected. The delay is not eliminable either.
 
-* rule_id
-* target ID
-* target name
-* constraints
-* role
-* replicas
-* scheduling state
+All the approaches above will result in a delay. Fortunately, for case 1 and 2 above, delay is acceptable. It doesn’t matter much if the optimizer doesn’t perceive the placement rules changement immediately. The worst result is that the latency is relatively high for a short time.
 
-The system table is a virtual table, which doesn’t persist data. When querying the table, TiDB queries PD and integrates the result in a table format. That also means the metadata is stored on PD instead of TiKV.
+For case 3, although delay is acceptable, but all TiDB instances must be always consistent on the placement rules. To achieve this goal, the schema version needs to be updated, thus transactions with old placement rules will fail when committed.
```

Comment:
My understanding is that the placement meta data is required to be correct so that local transactions can function correctly. i.e. if the location has been updated, the transaction might not be able to be function as a local trx any more.

Reply:
Had discussed with ailinKid, just a note: the failure comes from schema version. 2pc/1pc/async check if they should fail by themselves. And, in my understanding, you can't strictly force the transaction to be updated with placement rules updated.